bbbc021_feature_reduction_tests.py

```diff
@@ -0,0 +1,247 @@
+
+    :param filenames: list of valid filenames with relative paths from top level directory
+    :param names_out: list of valid filenames for saving output with relative paths from top level directory
+    """
+    illum = find_background_illumination(filenames)
+    corrected_images = correct_multiimage_illumination(filenames, illum=illum)
+    for (image, name) in zip(corrected_images, names_out):
+        io.imsave(name, image)
+
+
+def run_montage(filenames, names_out):
+    """
+    Read images from filenames and stitch and stack their quadrants and channels before saving to new files using
+    names_out
+
+    :param filenames: list of filenames with relative paths to top level sorted by well, quadrant and channel e.g.
+                        filenames = ['B02_s1_w1_illum.tif', 'B02_s1_w2_illum.tif', 'B02_s1_w4_illum.tif',
+                                    'B02_s2_w1_illum.tif', 'B02_s2_w2_illum.tif', 'B02_s2_w4_illum.tif',
+                                    'B02_s3_w1_illum.tif', 'B02_s3_w2_illum.tif', 'B02_s3_w4_illum.tif',
+                                    'B02_s4_w1_illum.tif', 'B02_s4_w2_illum.tif', 'B02_s4_w4_illum.tif']
+                        will result in one image (B02) with quadrants [[s1, s2], [s3, s4]] where each quadrant
+                        is stacked in the order [w4, w2, w1]. This example assumes files at the top level directory
+    :param names_out: list of filenames with relative paths to top level for output
+    """
+    illumed_ims = map(io.imread, filenames)
+    montaged_ims = montage_stream(illumed_ims, montage_order=[[0, 1], [2, 3]], channel_order=[2, 1, 0])
+    for (image, name) in zip(montaged_ims, names_out):
+        io.imsave(name, image)
+
+
+def output_features(ims, filenames, out_file):
+    """
+    Build a default feature map for each image in ims and output a dataframe of
+    [filenames, features] to out_file as csv for reading in
+
+    :param ims: opened nparray images
+    :param filenames: filenames corresponding to each image in ims with relative path to top level directory
+    :param out_file: name of CSV file to save dataframe, with relative path to top level directory
+    """
+    # generate filenames column to exist as first column of feature DF
+    filenames_col = ["Filenames"]
+    filenames_col.extend(filenames)
+    filenames_col = pd.DataFrame(filenames_col)
+
+    all_image_features = pd.DataFrame()
+    # set up flag to only add header row once
+    flag = True
+    i = 1
+    for im, im_name in zip(ims, filenames):
+        image_features, feature_names = default_feature_map(im, sample_size=100)
+        # make sure header row is added to dataframe in first iteration
+        if flag:
+            all_image_features = all_image_features.append(pd.DataFrame(feature_names).transpose())
+            flag = False
+        image_features = pd.DataFrame(image_features).transpose()
```

Comment:
So I think this is probably slowing you down: appending to a dataframe is slow. You should keep the header from the first image, and then append feature arrays to a list. Then, only at the very end, concatenate all the feature arrays into a big one (using `np.stack(arrays_list)`), and make a pandas dataframe using the big array and the column names.

An even better alternative is to never make the dataframe in the first place: open a file, write the comma-separated header, then write the values of each row to the file. That way, if the project crashes, you don't have to start from scratch again!

Reply:
I can also attest that appending to dataframes is very slow. I didn't pick up on this as a potential issue when I looked at the code before, sorry. I recently wrote a script that ran *six times faster* when I switched appending to a dataframe to appending to lists instead.

Please use the "with open" method if you are going to write to file, shown here as best practise in the little book of python anitpatterns: https://docs.quantifiedcode.com/python-anti-patterns/maintainability/not_using_with_to_open_files.html?highlight=open